src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+import asyncio
+import sys
+import time
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar, Union, cast
+
+if sys.version_info < (3, 10):
+    from typing_extensions import ParamSpec
+else:
+    from typing import ParamSpec
+
+Numeric = Union[int, float]
+P = ParamSpec("P")
+T = TypeVar("T")
+A = TypeVar("A", bound=Callable[..., Awaitable[Any]])
+
+
+class UnavailableTokensError(Exception):
+    pass
+
+
+class TokenLimiter:
+    """
+    A simple in-memory rate-limiter implemented using the leaky bucket algorithm.
```

Comment:
```suggestion
    A in-memory rate-limiter implemented using the leaky bucket algorithm.
```
You can provide a link to the algoritm for clarity

Reply:
[Token bucket](https://en.wikipedia.org/wiki/Token_bucket#Algorithm) seems more precise, because there are two variants of the leaky bucket algorithm.

Reply:
I'm went with `TokenRateLimiter` and described how the metaphor differs from a standard leaky bucket in the docstring

Reply:
What I tried to say was that "token bucket" is the more precise name of the algorithm being employed here and since it already has its own Wikipedia [page](https://en.wikipedia.org/wiki/Token_bucket#Algorithm), we should just link to that instead of the one for "leaky bucket".